flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java

```diff
@@ -320,6 +326,171 @@ public void insertRealtimeData(@PathVariable("id") String assetID,
       realtimeDataRepository.save(data);
    }
 
+   public Map<String,Set<String>> getMetricAndAssetIdMap (Asset asset) {
+      Map<String,Set<String>> metricNameAndAssetIdsMap = new HashMap<String,Set<String>>();
+      Map<String, Map<String, Map<String, String>>> formulars = asset.getMetricsformulars();
+      if(formulars == null || formulars.isEmpty()) {
+         return metricNameAndAssetIdsMap;
+      }
+      Map<String, Map<String, String>> sensorFormulars = formulars.get(FlowgateConstant.SENSOR);
+      if(sensorFormulars == null || sensorFormulars.isEmpty()) {
+         return metricNameAndAssetIdsMap;
+      }
+      for(Map.Entry<String, Map<String, String>> sensorFormularMap : sensorFormulars.entrySet()) {
+         Map<String, String> sensorLocationAndIdMap = sensorFormularMap.getValue();
+         Set<String> assetIds = new HashSet<String>();
+         for(String formular : sensorLocationAndIdMap.values()) {
+            assetIds.addAll(Arrays.asList(formular.split("\\+|-|\\*|/|\\(|\\)")));
+         }
+         switch (sensorFormularMap.getKey()) {
+         case MetricName.PDU_HUMIDITY:
+            Set<String> humidityAssetIds = metricNameAndAssetIdsMap.get(MetricName.PDU_HUMIDITY);
+            if(humidityAssetIds == null) {
+               humidityAssetIds = new HashSet<String>();
+            }
+            humidityAssetIds.addAll(assetIds);
+            metricNameAndAssetIdsMap.put(MetricName.PDU_HUMIDITY, humidityAssetIds);
+            break;
+         case MetricName.PDU_TEMPERATURE:
+            Set<String> tempAssetIds = metricNameAndAssetIdsMap.get(MetricName.PDU_TEMPERATURE);
+            if(tempAssetIds == null) {
+               tempAssetIds = new HashSet<String>();
+            }
+            tempAssetIds.addAll(assetIds);
+            metricNameAndAssetIdsMap.put(MetricName.PDU_TEMPERATURE, tempAssetIds);
+            break;
+         default:
+            break;
+         }
+      }
+      return metricNameAndAssetIdsMap;
+   }
+
+   //Descending by time
+   public void realTimeDataSortByCreateTime(List<RealTimeData> realtimeDatas) {
+      Collections.sort(realtimeDatas, new Comparator<RealTimeData>() {
+         @Override
+         public int compare(RealTimeData data1, RealTimeData data2) {
+            if (data2.getTime() > data1.getTime()) {
+               return 1;
+            }
+            if (data2.getTime() == data1.getTime()) {
+               return 0;
+            }
+            return -1;
+         }
+      });
+   }
+
+   public List<ValueUnit> getValueUnits(List<RealTimeData> realtimeDatas){
+      List<ValueUnit> valueunits = null;
+      if(realtimeDatas == null || realtimeDatas.isEmpty()) {
+         return valueunits;
+      }
+      valueunits = new ArrayList<>();
+      realTimeDataSortByCreateTime(realtimeDatas);
```

Comment:
let;s consider use bubble , only fetch the largest.

Reply:
Done